flocker/node/agents/blockdevice.py

```diff
@@ -1264,6 +1381,72 @@ def shared_state_changes(self):
         )
 
 
+# Mapping from desired and discovered dataset state to
+# IStateChange factory. (The factory is expected to take
+# ``desired_dataset`` and ``discovered_dataset``.
+DATASET_TRANSITIONS = {
+    DatasetStates.MOUNTED: {
+        DatasetStates.NON_EXISTANT: CreateBlockDeviceDataset.from_dataset,
+        DatasetStates.ATTACHED_ELSEWHERE: lambda **kwargs: NoOp(),
+        DatasetStates.NON_MANIFEST: AttachVolume.from_dataset,
+        DatasetStates.ATTACHED: MountBlockDevice.from_dataset,
+    },
+    DatasetStates.NON_MANIFEST: {
+        DatasetStates.NON_EXISTANT: lambda **kwargs: NoOp(),
+        DatasetStates.ATTACHED_ELSEWHERE: lambda **kwargs: NoOp(),
+        DatasetStates.ATTACHED: DetachVolume.from_dataset,
+        DatasetStates.MOUNTED: UnmountBlockDevice.from_dataset,
+    },
+    DatasetStates.DELETED: {
+        DatasetStates.ATTACHED_ELSEWHERE: lambda **kwargs: NoOp(),
+        DatasetStates.NON_MANIFEST: lambda **kwargs: NoOp(),
+        DatasetStates.ATTACHED: lambda **kwargs: NoOp(),
+        DatasetStates.MOUNTED: DestroyBlockDeviceDataset.from_dataset,
+    },
+}
+
+
+@implementer(ICalculater)
+class BlockDeviceCalculater(PClass):
+    """
+    XXX
+    """
+    def _calculate_dataset_change(self, discovered_dataset, desired_dataset):
+
+        # If the configuration doesn't know about a dataset,
+        # we detach it.
+        desired_state = (desired_dataset.state
+                         if desired_dataset is not None
+                         else DatasetStates.NON_MANIFEST)
+        # If we haven't discovered a dataset, then it is doesn't
+        # exist.
+        discovered_state = (discovered_dataset.state
+                            if discovered_dataset is not None
+                            else DatasetStates.NON_EXISTANT)
+        if desired_state != discovered_state:
+            return DATASET_TRANSITIONS[desired_state][discovered_state](
+                discovered_dataset=discovered_dataset,
+                desired_dataset=desired_dataset,
+            )
+        else:
+            return NoOp()
+
+    def calculate_changes_for_datasets(
+        self, discovered_datasets, desired_datasets
+    ):
+        actions = []
+        # If a dataset isn't in the configuration, we don't act on it.
+        for dataset_id in set(discovered_datasets) | set(desired_datasets):
```

Comment:
I agree, but I thought that having it on a single line was even clearer. :disappointed: